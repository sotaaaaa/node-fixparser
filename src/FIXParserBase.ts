import type { ProxyAgent } from 'proxy-agent';

import { Enums as EnumsCache } from './enums/Enums';
import { Field } from './fields/Field';
import { Fields as FieldsCache } from './fields/Fields';
import { Field as FieldType } from './fieldtypes/Field';
import { Message } from './message/Message';
import type { LogOptions } from './util/Logger';
import { DEFAULT_FIX_VERSION, RE_ESCAPE, RE_FIND, SOH, STRING_EQUALS } from './util/util';

export type Protocol = 'tcp' | 'ssl-tcp' | 'tls-tcp' | 'websocket';
export type ConnectionType = 'acceptor' | 'initiator';

export type Options = {
  /* Target hostname or IP address */
  host?: string;
  /* Target port */
  port?: number;
  /* Type of protocol */
  protocol?: Protocol;
  /* Assigned value used to identify firm sending message, SenderCompID tag 49 */
  sender?: string;
  /* Assigned value used to identify receiving firm, TargetCompID tag 56 */
  target?: string;
  /* Heartbeat interval in seconds */
  heartbeatIntervalSeconds?: number;
  /* Protocol version, defaults to "FIXT.1.1" used by BeginString tag 8 */
  fixVersion?: string;
  /* Private keys in PEM format */
  tlsKey?: unknown | null;
  /* Cert chains in PEM format */
  tlsCert?: unknown | null;
  /* Use Server Name Indication extension of TLS (requires FQDN hostname) */
  tlsUseSNI?: boolean;
  /* Skip Standard Input pipe of TLS socket */
  tlsSkipStdInPipe?: boolean | null;
  /* Disable logs */
  logging?: boolean;
  /* Log options */
  logOptions?: LogOptions;
  /* Use a proxy connection */
  proxy?: ProxyAgent | null;
  /* Called when the FIX connection is opened and ready */
  onReady?: () => void;
  /* Called when a message has been received */
  onMessage?: (message: Message) => void;
  /* Called when the FIX connection has been initiated */
  onOpen?: () => void;
  /* Called when the FIX connection has been closed */
  onClose?: () => void;
  /* Called if the FIX connection has failed */
  onError?: (error?: Error) => void;
};

/**
 * FIXParser base class.
 */
export class FIXParserBase {
  /* Protocol version, defaults to "FIXT.1.1" used by BeginString tag 8 */
  public fixVersion: string = DEFAULT_FIX_VERSION;
  /* Current Message instance being processed */
  public message: Message | null = null;
  /* Current FIX message tag=value pair */
  public messageTags: string[] = [];
  /* Current FIX message in string format */
  public messageString = '';
  /* Collection of field values from the FIX specification */
  public fields: FieldsCache = new FieldsCache();
  /* Collection of enum values from the FIX specification */
  public enums: EnumsCache = new EnumsCache();

  /**
   * Processes a message by splitting up the string into tags.
   */
  public processMessage(): void {
    const matches: RegExpExecArray | null = RE_FIND.exec(this.messageString);
    if (matches && matches.length === 2) {
      const stringData: string = this.messageString.replace(
        new RegExp(matches[1].replace(RE_ESCAPE, '\\$&'), 'g'),
        SOH
      );
      this.message?.setString(stringData);
      this.messageTags = stringData.split(SOH);
    } else {
      this.message = null;
      this.messageTags = [];
    }
  }

  /**
   * Processes fields of a message.
   */
  public processFields(): void {
    let tag: number;
    let value: string | number | null;
    let i = 0;
    let equalsOperator: number;
    let field: Field;

    for (i; i < this.messageTags.length - 1; i++) {
      equalsOperator = this.messageTags[i].indexOf(STRING_EQUALS);

      tag = Number(this.messageTags[i].substring(0, equalsOperator));
      value = this.messageTags[i].substring(equalsOperator + 1);

      field = new Field(tag, value);

      this.fields.processField(this.message!, field);
      this.enums.processEnum(field);

      if (field.tag === FieldType.BeginString) {
        this.message!.fixVersion = String(field.value);
      } else if (field.tag === FieldType.BodyLength) {
        this.message?.validateBodyLength(value);
      } else if (field.tag === FieldType.CheckSum) {
        this.message?.validateChecksum(value);
      }

      this.message?.addField(field);
    }
  }

  /**
   * Parses FIX message(s) string into Message class instances.
   */
  public parse(data: string): Message[] {
    let i = 0;

    const messageStrings: string[] = data ? data.split(/(?<!\d)8=FIX/) : [];
    const messages: Message[] = [];

    for (i; i < messageStrings.length; i++) {
      this.message = new Message(this.fixVersion);
      this.messageString = `8=FIX${messageStrings[i]}`;
      if (this.messageString.indexOf(SOH) > -1) {
        this.message.setString(this.messageString);
        this.messageTags = this.messageString.split(SOH);
      } else {
        this.processMessage();
      }

      if (this.message) {
        this.processFields();
        messages.push(this.message);
      }
    }

    return messages;
  }
}
