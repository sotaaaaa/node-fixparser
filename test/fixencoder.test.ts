import { EncryptMethod, FIXParser, Field, Fields, type Message, Messages } from '../src/FIXParser';

describe('FIXEncoder', () => {
    const fixParser: FIXParser = new FIXParser();

    describe('#encode: tag ordering', () => {
        const fixString: string = '8=FIXT.1.1|9=34|35=1|49=ABC|56=XYZ|34=1|112=reqId|10=230|';

        it('should have encoded the FIX message in correct order 1', () => {
            const message: Message = fixParser.createMessage(
                new Field(Fields.MsgType, Messages.TestRequest),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.MsgSeqNum, fixParser.getNextTargetMsgSeqNum()),
                new Field(Fields.TestReqID, 'reqId'),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
        });

        it('should have encoded the FIX message in correct order 2', () => {
            const message: Message = fixParser.createMessage(
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.MsgSeqNum, fixParser.getNextTargetMsgSeqNum()),
                new Field(Fields.TestReqID, 'reqId'),
                new Field(Fields.MsgType, Messages.TestRequest),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
        });
    });
    describe('#encode: Heartbeat', () => {
        const fixString: string = '8=FIXT.1.1|9=51|35=0|34=703|49=ABC|52=20100130-10:53:40.830|56=XYZ|10=073|';

        it('should have encoded the FIX message', () => {
            const message: Message = fixParser.createMessage(
                new Field(Fields.MsgType, Messages.Heartbeat),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(703)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 53, 40, 830)))),
                new Field(Fields.TargetCompID, 'XYZ'),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
            expect(message.encode()).toMatchSnapshot();
        });
    });
    describe('#encode: Logon', () => {
        const fixString: string =
            '8=FIXT.1.1|9=76|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=214|';

        it('should have encoded the FIX message', () => {
            const message: Message = fixParser.createMessage(
                new Field(Fields.MsgType, Messages.Logon),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(702)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 52, 40, 663)))),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.RawDataLength, 4),
                new Field(Fields.RawData, 1234),
                new Field(Fields.EncryptMethod, EncryptMethod.None),
                new Field(Fields.HeartBtInt, 60),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
            expect(message.encode()).toMatchSnapshot();
        });
    });
    describe('#encode: Custom header/trailer', () => {
        let fixString = null;

        it('should have encoded the FIX message with custom FIX version', () => {
            fixString =
                '8=FIX.4.3|9=76|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=135|';
            const message: Message = fixParser.createMessage(
                new Field(Fields.BeginString, 'FIX.4.3'),
                new Field(Fields.MsgType, Messages.Logon),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(702)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 52, 40, 663)))),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.RawDataLength, 4),
                new Field(Fields.RawData, 1234),
                new Field(Fields.EncryptMethod, EncryptMethod.None),
                new Field(Fields.HeartBtInt, 60),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
            expect(message.encode()).toMatchSnapshot();
        });

        it('should have encoded the FIX message with custom Fields.BodyLength', () => {
            fixString =
                '8=FIXT.1.1|9=123456|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=158|';
            const message: Message = fixParser.createMessage(
                new Field(Fields.MsgType, Messages.Logon),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(702)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 52, 40, 663)))),
                new Field(Fields.BodyLength, 123456),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.RawDataLength, 4),
                new Field(Fields.RawData, 1234),
                new Field(Fields.EncryptMethod, EncryptMethod.None),
                new Field(Fields.HeartBtInt, 60),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
            expect(message.encode()).toMatchSnapshot();
        });

        it('should have encoded the FIX message with custom Fields.CheckSum', () => {
            fixString =
                '8=FIXT.1.1|9=76|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=999|';
            const message: Message = fixParser.createMessage(
                new Field(Fields.CheckSum, 999),
                new Field(Fields.MsgType, Messages.Logon),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(702)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 52, 40, 663)))),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.RawDataLength, 4),
                new Field(Fields.RawData, 1234),
                new Field(Fields.EncryptMethod, EncryptMethod.None),
                new Field(Fields.HeartBtInt, 60),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
            expect(message.encode()).toMatchSnapshot();
        });
    });
    describe('#encode: Multiple encode', () => {
        let fixString =
            '8=FIX.4.3|9=76|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=135|';

        it('should have encoded the FIX message 1', () => {
            const message: Message = fixParser.createMessage(
                new Field(Fields.BeginString, 'FIX.4.3'),
                new Field(Fields.MsgType, Messages.Logon),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(702)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 52, 40, 663)))),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.RawDataLength, 4),
                new Field(Fields.RawData, 1234),
                new Field(Fields.EncryptMethod, EncryptMethod.None),
                new Field(Fields.HeartBtInt, 60),
            );
            const encodedMessage0: string = message.encode().replace(/\x01/g, '|');
            const encodedMessage1: string = message.encode();
            const encodedMessage2: string = message.encode();
            expect(encodedMessage0).toEqual(fixString);
            expect(encodedMessage1).toEqual(encodedMessage2);
            expect(encodedMessage0).toMatchSnapshot();
            expect(encodedMessage1).toMatchSnapshot();
        });

        it('should have encoded the FIX message 2', () => {
            fixString =
                '8=FIXT.1.1|9=76|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=214|';
            const message: Message = fixParser.createMessage(
                new Field(Fields.MsgType, Messages.Logon),
                new Field(Fields.MsgSeqNum, fixParser.setNextTargetMsgSeqNum(702)),
                new Field(Fields.SenderCompID, 'ABC'),
                new Field(Fields.SendingTime, fixParser.getTimestamp(new Date(Date.UTC(2010, 0, 30, 10, 52, 40, 663)))),
                new Field(Fields.TargetCompID, 'XYZ'),
                new Field(Fields.RawDataLength, 4),
                new Field(Fields.RawData, 1234),
                new Field(Fields.EncryptMethod, EncryptMethod.None),
                new Field(Fields.HeartBtInt, 60),
            );
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(fixParser.parse(message.encode())[0].messageString.replace(/\x01/g, '|')).toEqual(fixString);
            message.encode('-');
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(fixParser.parse(message.encode())[0].messageString.replace(/\x01/g, '|')).toEqual(fixString);
            message.encode('/');
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(fixParser.parse(message.encode())[0].messageString.replace(/\x01/g, '|')).toEqual(fixString);
            message.encode('_');
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(fixParser.parse(message.encode())[0].messageString.replace(/\x01/g, '|')).toEqual(fixString);
            message.encode('*');
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(fixParser.parse(message.encode())[0].messageString.replace(/\x01/g, '|')).toEqual(fixString);
            message.encode('$');
            expect(message.encode().replace(/\x01/g, '|')).toEqual(fixString);
            expect(message).toMatchSnapshot();
            expect(message.encode()).toMatchSnapshot();
        });
    });
});
