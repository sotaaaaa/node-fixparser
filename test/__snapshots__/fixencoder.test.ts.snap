// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FIXEncoder #encode: Custom header/trailer should have encoded the FIX message with custom FIX version 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "A",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 8,
      "type": null,
      "value": "FIX.4.3",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 702,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 52,
      "type": null,
      "value": "20100130-10:52:40.663",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 95,
      "type": null,
      "value": 4,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 96,
      "type": null,
      "value": 1234,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 98,
      "type": null,
      "value": 0,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 108,
      "type": null,
      "value": 60,
    },
  ],
  "description": "",
  "fixVersion": "FIX.4.3",
  "messageSequence": 702,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: Custom header/trailer should have encoded the FIX message with custom FIX version 2`] = `"8=FIX.4.39=7635=A34=70249=ABC52=20100130-10:52:40.66356=XYZ95=496=123498=0108=6010=135"`;

exports[`FIXEncoder #encode: Custom header/trailer should have encoded the FIX message with custom Fields.BodyLength 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "A",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 702,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 52,
      "type": null,
      "value": "20100130-10:52:40.663",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 9,
      "type": null,
      "value": 123456,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 95,
      "type": null,
      "value": 4,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 96,
      "type": null,
      "value": 1234,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 98,
      "type": null,
      "value": 0,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 108,
      "type": null,
      "value": 60,
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 702,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: Custom header/trailer should have encoded the FIX message with custom Fields.BodyLength 2`] = `"8=FIXT.1.19=12345635=A34=70249=ABC52=20100130-10:52:40.66356=XYZ95=496=123498=0108=6010=158"`;

exports[`FIXEncoder #encode: Custom header/trailer should have encoded the FIX message with custom Fields.CheckSum 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "A",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 10,
      "type": null,
      "value": 999,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 702,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 52,
      "type": null,
      "value": "20100130-10:52:40.663",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 95,
      "type": null,
      "value": 4,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 96,
      "type": null,
      "value": 1234,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 98,
      "type": null,
      "value": 0,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 108,
      "type": null,
      "value": 60,
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 702,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: Custom header/trailer should have encoded the FIX message with custom Fields.CheckSum 2`] = `"8=FIXT.1.19=7635=A34=70249=ABC52=20100130-10:52:40.66356=XYZ95=496=123498=0108=6010=999"`;

exports[`FIXEncoder #encode: Heartbeat should have encoded the FIX message 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "0",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 703,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 52,
      "type": null,
      "value": "20100130-10:53:40.830",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 703,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: Heartbeat should have encoded the FIX message 2`] = `"8=FIXT.1.19=5135=034=70349=ABC52=20100130-10:53:40.83056=XYZ10=073"`;

exports[`FIXEncoder #encode: Logon should have encoded the FIX message 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "A",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 702,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 52,
      "type": null,
      "value": "20100130-10:52:40.663",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 95,
      "type": null,
      "value": 4,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 96,
      "type": null,
      "value": 1234,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 98,
      "type": null,
      "value": 0,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 108,
      "type": null,
      "value": 60,
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 702,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: Logon should have encoded the FIX message 2`] = `"8=FIXT.1.19=7635=A34=70249=ABC52=20100130-10:52:40.66356=XYZ95=496=123498=0108=6010=214"`;

exports[`FIXEncoder #encode: Multiple encode should have encoded the FIX message 1 1`] = `"8=FIX.4.3|9=76|35=A|34=702|49=ABC|52=20100130-10:52:40.663|56=XYZ|95=4|96=1234|98=0|108=60|10=135|"`;

exports[`FIXEncoder #encode: Multiple encode should have encoded the FIX message 1 2`] = `"8=FIX.4.39=7635=A34=70249=ABC52=20100130-10:52:40.66356=XYZ95=496=123498=0108=6010=135"`;

exports[`FIXEncoder #encode: Multiple encode should have encoded the FIX message 2 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "A",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 702,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 52,
      "type": null,
      "value": "20100130-10:52:40.663",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 95,
      "type": null,
      "value": 4,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 96,
      "type": null,
      "value": 1234,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 98,
      "type": null,
      "value": 0,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 108,
      "type": null,
      "value": 60,
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 702,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: Multiple encode should have encoded the FIX message 2 2`] = `"8=FIXT.1.19=7635=A34=70249=ABC52=20100130-10:52:40.66356=XYZ95=496=123498=0108=6010=214"`;

exports[`FIXEncoder #encode: tag ordering should have encoded the FIX message in correct order 1 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "1",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 1,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 112,
      "type": null,
      "value": "reqId",
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 1,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;

exports[`FIXEncoder #encode: tag ordering should have encoded the FIX message in correct order 2 1`] = `
Message {
  "bodyLengthExpected": null,
  "bodyLengthValid": false,
  "bodyLengthValue": null,
  "checksumExpected": null,
  "checksumValid": false,
  "checksumValue": null,
  "data": [
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 35,
      "type": null,
      "value": "1",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 49,
      "type": null,
      "value": "ABC",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 56,
      "type": null,
      "value": "XYZ",
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 34,
      "type": null,
      "value": 1,
    },
    Field {
      "category": null,
      "description": null,
      "enumeration": null,
      "name": null,
      "section": null,
      "tag": 112,
      "type": null,
      "value": "reqId",
    },
  ],
  "description": "",
  "fixVersion": "FIXT.1.1",
  "messageSequence": 1,
  "messageString": "",
  "messageType": "",
  "messageTypeDescription": "",
  "structure": null,
}
`;
